### PR TITLE
Added first-pass declaration file for Typescript.

### PR DESCRIPTION
**Pull request by aaronsky**
This PR partially addresses #192. I have scanned the Slack and core pieces of botkit for necessary declaration matter and I wanted to get a discussion going around including this. Bundling the declaration file makes use with TypeScript much easier than going through DefinitelyTyped, so hopefully I can get some feedback on this from people and maybe make some progress on the not-Slack parts of the codebase.

### CODE REVIEW

**Comment by MattJeanes** [comment]
This looks great, one of the only things that was putting me off from using BotKit was the lack of TypeScript declarations - I still use it but I ended up writing something similar but a lot more lightweight than this. I really hope this gets officially included soon! 👍

**Comment by theluk** [comment]
is there any reason why this is not included?

File: botkit.d.ts [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+        on(event: string, cb: Function): never;
+        trigger(event: string, data): void;
+        /**
+         * proceed to the next message after waiting for an answer
+         */
+        next(): never;
+        repeat(): never;
+        silentRepeat(): never;
+        addQuestion(message: Message, cb: Function, capture_options, thread): never;
+        ask(message: Message, cb: Function, capture_options): never;
```
Comment (mtgto) [inline botkit.d.ts]: Using this type definition, found some odd types in Conversation class:
- `cb` might be either `Function` or `array of mapping of keyword to callback`.
- `message` might be either `string` or `Message`.
See: https://github.com/howdyai/botkit#using-conversationask-with-an-array-of-callbacks
Example: https://github.com/howdyai/botkit#using-conversationask-with-an-array-of-callbacks

It works for my project:
```diff
diff --git a/botkit.d.ts b/botkit.d.ts
index eac474b..ffd615a 100644
--- a/botkit.d.ts
+++ b/botkit.d.ts
@@ -373,8 +373,8 @@ declare module botkit {
         next(): never;
         repeat(): never;
         silentRepeat(): never;
-        addQuestion(message: Message, cb: Function, capture_options, thread): never;
-        ask(message: Message, cb: Function, capture_options): never;
+        addQuestion(message: string | Message, cb: Function | {pattern?: string | RegExp, callback: Function, default?: boolean}[], capture_options, thread): never;
+        ask(message: string | Message, cb: Function | {pattern?: string | RegExp, callback: Function, default?: boolean}[], capture_options): never;
         addMessage(message: Message, thread): never;
         /**
          * how long should the bot wait while a user answers?
```

**Comment by cesarvarela** [comment]
News about this? How can we help

**Comment by r-arias** [comment]
I'm definitely very interested in this being included.

File: botkit.d.ts [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+        closeRTM(err: Error): never;
+        reconnect(err: Error): any;
+        destroy(): never;
+        startRTM(cb: (err: string | Error, bot: Bot, res) => void): Bot;
+        identifyBot(cb: (err: string | Error, data) => void): any;
+        identifyTeam(cb: (err: string | Error, id: string) => void): string;
+        startPrivateConversation(message: Message, cb: (err: Error, convo: Conversation) => void): never;
+        startConversation(message: Message, cb: (err: Error, convo: Conversation) => void): never;
+        createConversation(message: Message, cb: (err: Error, convo: Conversation) => void): never;
+        private _startDM(task: Task, user_id: string, cb: Function): any;
```
Comment (mmmveggies) [inline botkit.d.ts]: Any reason this was prefixed with `private` ?
Although the `_` prefix does seem to imply that, [at implementation](https://github.com/howdyai/botkit/tree/master/lib/Slackbot_worker.js#L326-L338) there's no evidence that it should be enforced or that we should break usages of the method in the wild.

File: botkit.d.ts [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+        replyPublic(src: any, resp: string, cb: (err?: string | Error) => void): never;
+        replyPublicDelayed(src: any, resp: string, cb: (err?: string | Error) => void): never;
+        replyPrivate(src: any, resp: string, cb: (err?: string | Error) => void): never;
+        replyPrivateDelayed(src: any, resp: string, cb: (err?: string | Error) => void): never;
+        replyInteractive(src: any, resp: string, cb: (err?: string | Error) => void): never;
+        reply(src: Message, resp: string);
+        reply(src: Message, resp: string, cb?: (err?: string | Error) => void): never;
+        say(message: Message, cb?: (err?: string | Error) => void): never;
+        startTyping(src: any): never;
+        replyWithTyping(src: any, resp: string, cb: (err?: string | Error) => void): never;
```
Comment (indierojo) [inline botkit.d.ts]: cb is optional here -- and probably most places

File: botkit.d.ts [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+        startRTM(cb: (err: string | Error, bot: Bot, res) => void): Bot;
+        identifyBot(cb: (err: string | Error, data) => void): any;
+        identifyTeam(cb: (err: string | Error, id: string) => void): string;
+        startPrivateConversation(message: Message, cb: (err: Error, convo: Conversation) => void): never;
+        startConversation(message: Message, cb: (err: Error, convo: Conversation) => void): never;
+        createConversation(message: Message, cb: (err: Error, convo: Conversation) => void): never;
+        private _startDM(task: Task, user_id: string, cb: Function): any;
+        send(message: Message, cb: (err?: string | Error) => void): never;
+        replyAcknowledge(cb: (err?: string | Error) => void): never;
+        replyPublic(src: any, resp: string, cb: (err?: string | Error) => void): never;
```
Comment (indierojo) [inline botkit.d.ts]: All of the reply methods also support a different reply type:
```
    interface Reply {
        username: string,
        attachments: Array<Attachment>
    }
```
`resp: string` should be `resp: string | Reply` for each of the following:

- replyPublic
- replyPublicDelayed
- replyPrivate
- replyPrivateDelayed
- replyInteractive
- reply
- replyWithTyping
- replyAndUpdate

**Comment by pascalwhoop** [comment]
@aaronsky can you update the PR with the changes mentioned by @indierojo ? Or shall I clone your repo and make another PR?

**Comment by Naktibalda** [comment]
This PR was superseded by #953, please close it.

**Pull request closed without merging**